Error when adding or removing a user from an organisation when there are 2 or more Keycloak servers in a cluster

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

organizations

### Describe the bug

If I add or remove a user from an organization I am getting the error below - note that this only occurs in a cluster where there is more than 1 server:

Caused by: java.lang.ClassCastException: class org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent cannot be cast to class org.keycloak.models.cache.infinispan.events.UserCacheInvalidationEvent (org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent and org.keycloak.models.cache.infinispan.events.UserCacheInvalidationEvent are in unnamed module of loader io.quarkus.bootstrap.runner.RunnerClassLoader @7e6cbb7a)

### Version

26.0.0

### Regression

- [ ] The issue is a regression

### Expected behavior

I would expect to add and remove a user from an organization without errors occuring.

### Actual behavior

When I add or remove a user from an organization an error occurs - although the user does actually seem to be added / removed, leading me to believe it has something to do with the caching rather than the actual mechanics of adding or removing the user.

### How to Reproduce?

Run a 2 Keycloak server cluster. I am using kubernetes and the codecentric/keycloakx helm chart found here https://codecentric.github.io/helm-charts. 

Create a realm and enable organizations.
Add an organization to the realm
Add a test user
Add the user to the organization

### Anything else?

_No response_

Hi @pedroigor - thanks for responding so quickly! I have pasted the stack trace below. Please let me know if I can help in anyway. I'm not sure if this is some configuration issue in my helm chart, but everything else seems to be working happily.


2024-10-10 11:36:25,594 ERROR [org.infinispan.interceptors.impl.InvocationContextInterceptor] (jgroups-15,keycloak26-keycloakx-0-47006) ISPN000136: Error executing command PutKeyValueCommand on Cache 'work', writing keys [4ea2ba7b-40ad-41ab-aaaa-be626c2ff214]: org.infinispan.commons.CacheListenerException: ISPN000280: Caught exception [java.lang.ClassCastException] while invoking method [public void org.keycloak.cluster.infinispan.InfinispanNotificationsManager$CacheEntryListener.cacheEntryCreated(org.infinispan.notifications.cachelistener.event.CacheEntryCreatedEvent)] on listener instance: org.keycloak.cluster.infinispan.InfinispanNotificationsManager$CacheEntryListener@1e03f0cb
        at org.infinispan.notifications.impl.AbstractListenerImpl$ListenerInvocationImpl.lambda$invoke$1(AbstractListenerImpl.java:437)
        at org.infinispan.notifications.impl.AbstractListenerImpl$ListenerInvocationImpl.invoke(AbstractListenerImpl.java:457)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.doRealInvocation(CacheNotifierImpl.java:1793)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.doRealInvocation(CacheNotifierImpl.java:1789)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.invokeNoChecks(CacheNotifierImpl.java:1784)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.invoke(CacheNotifierImpl.java:1757)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl.doNotifyCreated(CacheNotifierImpl.java:428)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl.notifyCacheEntryCreated(CacheNotifierImpl.java:410)
        at org.infinispan.notifications.cachelistener.NotifyHelper.entryCommitted(NotifyHelper.java:63)
        at org.infinispan.interceptors.locking.ClusteringDependentLogic$ReplicationLogic.commitSingleEntry(ClusteringDependentLogic.java:505)
        at org.infinispan.interceptors.locking.ClusteringDependentLogic$AbstractClusteringDependentLogic.commitEntry(ClusteringDependentLogic.java:239)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.commitContextEntry(EntryWrappingInterceptor.java:611)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.commitEntryIfNeeded(EntryWrappingInterceptor.java:884)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.commitContextEntries(EntryWrappingInterceptor.java:591)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.applyChanges(EntryWrappingInterceptor.java:685)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.applyAndFixVersion(EntryWrappingInterceptor.java:737)
        at org.infinispan.interceptors.SyncInvocationStage.thenApply(SyncInvocationStage.java:44)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.setSkipRemoteGetsAndInvokeNextForDataCommand(EntryWrappingInterceptor.java:732)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.visitPutKeyValueCommand(EntryWrappingInterceptor.java:335)
        at org.infinispan.commands.write.PutKeyValueCommand.acceptVisitor(PutKeyValueCommand.java:65)
        at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNext(BaseAsyncInterceptor.java:58)
        at org.infinispan.interceptors.locking.AbstractLockingInterceptor.visitNonTxDataWriteCommand(AbstractLockingInterceptor.java:134)
        at org.infinispan.interceptors.locking.NonTransactionalLockingInterceptor.visitDataWriteCommand(NonTransactionalLockingInterceptor.java:41)
        at org.infinispan.interceptors.locking.AbstractLockingInterceptor.visitPutKeyValueCommand(AbstractLockingInterceptor.java:82)
        at org.infinispan.commands.write.PutKeyValueCommand.acceptVisitor(PutKeyValueCommand.java:65)
        at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNext(BaseAsyncInterceptor.java:58)
        at org.infinispan.statetransfer.StateTransferInterceptor.handleNonTxWriteCommand(StateTransferInterceptor.java:309)
        at org.infinispan.statetransfer.StateTransferInterceptor.handleWriteCommand(StateTransferInterceptor.java:256)
        at org.infinispan.statetransfer.StateTransferInterceptor.visitPutKeyValueCommand(StateTransferInterceptor.java:96)
        at org.infinispan.commands.write.PutKeyValueCommand.acceptVisitor(PutKeyValueCommand.java:65)
        at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNext(BaseAsyncInterceptor.java:58)
        at org.infinispan.interceptors.impl.CacheMgmtInterceptor.updateStoreStatistics(CacheMgmtInterceptor.java:233)
        at org.infinispan.interceptors.impl.CacheMgmtInterceptor.visitPutKeyValueCommand(CacheMgmtInterceptor.java:196)
        at org.infinispan.commands.write.PutKeyValueCommand.acceptVisitor(PutKeyValueCommand.java:65)
        at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNext(BaseAsyncInterceptor.java:58)
        at org.infinispan.interceptors.DDAsyncInterceptor.handleDefault(DDAsyncInterceptor.java:54)
        at org.infinispan.interceptors.DDAsyncInterceptor.visitPutKeyValueCommand(DDAsyncInterceptor.java:60)
        at org.infinispan.commands.write.PutKeyValueCommand.acceptVisitor(PutKeyValueCommand.java:65)
        at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNextAndExceptionally(BaseAsyncInterceptor.java:127)
        at org.infinispan.interceptors.impl.InvocationContextInterceptor.visitCommand(InvocationContextInterceptor.java:90)
        at org.infinispan.interceptors.impl.AsyncInterceptorChainImpl.invokeAsync(AsyncInterceptorChainImpl.java:211)
        at org.infinispan.commands.remote.SingleRpcCommand.invokeAsync(SingleRpcCommand.java:72)
        at org.infinispan.remoting.inboundhandler.BasePerCacheInboundInvocationHandler.invokeCommand(BasePerCacheInboundInvocationHandler.java:108)
        at org.infinispan.remoting.inboundhandler.BaseBlockingRunnable.invoke(BaseBlockingRunnable.java:111)
        at org.infinispan.remoting.inboundhandler.BaseBlockingRunnable.runAsync(BaseBlockingRunnable.java:79)
        at org.infinispan.remoting.inboundhandler.BaseBlockingRunnable.run(BaseBlockingRunnable.java:48)
        at org.infinispan.remoting.inboundhandler.BasePerCacheInboundInvocationHandler.handleRunnable(BasePerCacheInboundInvocationHandler.java:154)
        at org.infinispan.remoting.inboundhandler.NonTxPerCacheInboundInvocationHandler.handle(NonTxPerCacheInboundInvocationHandler.java:44)
        at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler.handleCacheRpcCommand(GlobalInboundInvocationHandler.java:123)
        at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler.handleFromCluster(GlobalInboundInvocationHandler.java:77)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processRequest(JGroupsTransport.java:1527)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1454)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1649)
        at org.jgroups.JChannel.up(JChannel.java:748)
        at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:941)
        at org.jgroups.protocols.FRAG2.up(FRAG2.java:139)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:261)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:253)
        at org.jgroups.protocols.pbcast.GMS.up(GMS.java:853)
        at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:235)
        at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:462)
        at org.jgroups.protocols.pbcast.NAKACK2.deliver(NAKACK2.java:1006)
        at org.jgroups.protocols.pbcast.NAKACK2.handleMessage(NAKACK2.java:894)
        at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:679)
        at org.jgroups.protocols.VERIFY_SUSPECT2.up(VERIFY_SUSPECT2.java:105)
        at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:180)
        at org.jgroups.protocols.FD_SOCK2.up(FD_SOCK2.java:190)
        at org.jgroups.protocols.MERGE3.up(MERGE3.java:274)
        at org.jgroups.protocols.Discovery.up(Discovery.java:296)
        at org.jgroups.stack.Protocol.up(Protocol.java:360)
        at org.jgroups.protocols.TP.passMessageUp(TP.java:1230)
        at org.jgroups.util.SubmitToThreadPool$SingleMessageHandler.run(SubmitToThreadPool.java:95)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.ClassCastException: class org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent cannot be cast to class org.keycloak.models.cache.infinispan.events.UserCacheInvalidationEvent (org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent and org.keycloak.models.cache.infinispan.events.UserCacheInvalidationEvent are in unnamed module of loader io.quarkus.bootstrap.runner.RunnerClassLoader @7e6cbb7a)
        at org.keycloak.models.cache.infinispan.UserCacheManager.addInvalidationsFromEvent(UserCacheManager.java:96)
        at org.keycloak.models.cache.infinispan.CacheManager.invalidationEventReceived(CacheManager.java:218)
        at org.keycloak.models.cache.infinispan.InfinispanUserCacheProviderFactory.lambda$lazyInit$0(InfinispanUserCacheProviderFactory.java:65)
        at org.keycloak.cluster.ClusterEvent.accept(ClusterEvent.java:29)
        at org.keycloak.cluster.ClusterEvent.accept(ClusterEvent.java:25)
        at java.base/java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891)
        at org.keycloak.cluster.infinispan.InfinispanNotificationsManager.eventReceived(InfinispanNotificationsManager.java:290)
        at org.keycloak.cluster.infinispan.InfinispanNotificationsManager$CacheEntryListener.cacheEntryCreated(InfinispanNotificationsManager.java:187)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.infinispan.notifications.impl.AbstractListenerImpl$ListenerInvocationImpl.lambda$invoke$1(AbstractListenerImpl.java:431)
        ... 74 more

2024-10-10 11:36:25,594 WARN  [org.infinispan.CLUSTER] (jgroups-15,keycloak26-keycloakx-0-47006) ISPN000071: Caught exception when handling command SingleRpcCommand{cacheName='work', command=PutKeyValueCommand{key=4ea2ba7b-40ad-41ab-aaaa-be626c2ff214, value=WrapperClusterEvent [ eventKey=USER_INVALIDATION_EVENTS, sender=keycloak26-keycloakx-1-37804, senderSite=null, delegateEvents=[UserUpdatedEvent [ userId=572eec66-c00f-4c89-91b8-d84997e27835, username=matthew@mapli.co.za, email=matthew@mapli.co.za ], org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent@bf1178c6, org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent@2a2fce4f] ], flags=[SKIP_CACHE_STORE, IGNORE_RETURN_VALUES, BACKUP_WRITE], commandInvocationId=CommandInvocation:keycloak26-keycloakx-1-37804:20, putIfAbsent=false, returnEntry=false, valueMatcher=MATCH_ALWAYS, metadata=EmbeddedExpirableMetadata{version=null, lifespan=120000, maxIdle=-1}, internalMetadata=null, successful=true, topologyId=15, segment=10}}: org.infinispan.commons.CacheListenerException: ISPN000280: Caught exception [java.lang.ClassCastException] while invoking method [public void org.keycloak.cluster.infinispan.InfinispanNotificationsManager$CacheEntryListener.cacheEntryCreated(org.infinispan.notifications.cachelistener.event.CacheEntryCreatedEvent)] on listener instance: org.keycloak.cluster.infinispan.InfinispanNotificationsManager$CacheEntryListener@1e03f0cb
        at org.infinispan.notifications.impl.AbstractListenerImpl$ListenerInvocationImpl.lambda$invoke$1(AbstractListenerImpl.java:437)
        at org.infinispan.notifications.impl.AbstractListenerImpl$ListenerInvocationImpl.invoke(AbstractListenerImpl.java:457)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.doRealInvocation(CacheNotifierImpl.java:1793)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.doRealInvocation(CacheNotifierImpl.java:1789)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.invokeNoChecks(CacheNotifierImpl.java:1784)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.invoke(CacheNotifierImpl.java:1757)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl.doNotifyCreated(CacheNotifierImpl.java:428)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl.notifyCacheEntryCreated(CacheNotifierImpl.java:410)
        at org.infinispan.notifications.cachelistener.NotifyHelper.entryCommitted(NotifyHelper.java:63)
        at org.infinispan.interceptors.locking.ClusteringDependentLogic$ReplicationLogic.commitSingleEntry(ClusteringDependentLogic.java:505)
        at org.infinispan.interceptors.locking.ClusteringDependentLogic$AbstractClusteringDependentLogic.commitEntry(ClusteringDependentLogic.java:239)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.commitContextEntry(EntryWrappingInterceptor.java:611)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.commitEntryIfNeeded(EntryWrappingInterceptor.java:884)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.commitContextEntries(EntryWrappingInterceptor.java:591)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.applyChanges(EntryWrappingInterceptor.java:685)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.applyAndFixVersion(EntryWrappingInterceptor.java:737)
        at org.infinispan.interceptors.SyncInvocationStage.thenApply(SyncInvocationStage.java:44)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.setSkipRemoteGetsAndInvokeNextForDataCommand(EntryWrappingInterceptor.java:732)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.visitPutKeyValueCommand(EntryWrappingInterceptor.java:335)
        at org.infinispan.commands.write.PutKeyValueCommand.acceptVisitor(PutKeyValueCommand.java:65)
        at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNext(BaseAsyncInterceptor.java:58)
        at org.infinispan.interceptors.locking.AbstractLockingInterceptor.visitNonTxDataWriteCommand(AbstractLockingInterceptor.java:134)
        at org.infinispan.interceptors.locking.NonTransactionalLockingInterceptor.visitDataWriteCommand(NonTransactionalLockingInterceptor.java:41)
        at org.infinispan.interceptors.locking.AbstractLockingInterceptor.visitPutKeyValueCommand(AbstractLockingInterceptor.java:82)
        at org.infinispan.commands.write.PutKeyValueCommand.acceptVisitor(PutKeyValueCommand.java:65)
        at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNext(BaseAsyncInterceptor.java:58)
        at org.infinispan.statetransfer.StateTransferInterceptor.handleNonTxWriteCommand(StateTransferInterceptor.java:309)
        at org.infinispan.statetransfer.StateTransferInterceptor.handleWriteCommand(StateTransferInterceptor.java:256)
        at org.infinispan.statetransfer.StateTransferInterceptor.visitPutKeyValueCommand(StateTransferInterceptor.java:96)
        at org.infinispan.commands.write.PutKeyValueCommand.acceptVisitor(PutKeyValueCommand.java:65)
        at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNext(BaseAsyncInterceptor.java:58)
        at org.infinispan.interceptors.impl.CacheMgmtInterceptor.updateStoreStatistics(CacheMgmtInterceptor.java:233)
        at org.infinispan.interceptors.impl.CacheMgmtInterceptor.visitPutKeyValueCommand(CacheMgmtInterceptor.java:196)
        at org.infinispan.commands.write.PutKeyValueCommand.acceptVisitor(PutKeyValueCommand.java:65)
        at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNext(BaseAsyncInterceptor.java:58)
        at org.infinispan.interceptors.DDAsyncInterceptor.handleDefault(DDAsyncInterceptor.java:54)
        at org.infinispan.interceptors.DDAsyncInterceptor.visitPutKeyValueCommand(DDAsyncInterceptor.java:60)
        at org.infinispan.commands.write.PutKeyValueCommand.acceptVisitor(PutKeyValueCommand.java:65)
        at org.infinispan.interceptors.BaseAsyncInterceptor.invokeNextAndExceptionally(BaseAsyncInterceptor.java:127)
        at org.infinispan.interceptors.impl.InvocationContextInterceptor.visitCommand(InvocationContextInterceptor.java:90)
        at org.infinispan.interceptors.impl.AsyncInterceptorChainImpl.invokeAsync(AsyncInterceptorChainImpl.java:211)
        at org.infinispan.commands.remote.SingleRpcCommand.invokeAsync(SingleRpcCommand.java:72)
        at org.infinispan.remoting.inboundhandler.BasePerCacheInboundInvocationHandler.invokeCommand(BasePerCacheInboundInvocationHandler.java:108)
        at org.infinispan.remoting.inboundhandler.BaseBlockingRunnable.invoke(BaseBlockingRunnable.java:111)
        at org.infinispan.remoting.inboundhandler.BaseBlockingRunnable.runAsync(BaseBlockingRunnable.java:79)
        at org.infinispan.remoting.inboundhandler.BaseBlockingRunnable.run(BaseBlockingRunnable.java:48)
        at org.infinispan.remoting.inboundhandler.BasePerCacheInboundInvocationHandler.handleRunnable(BasePerCacheInboundInvocationHandler.java:154)
        at org.infinispan.remoting.inboundhandler.NonTxPerCacheInboundInvocationHandler.handle(NonTxPerCacheInboundInvocationHandler.java:44)
        at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler.handleCacheRpcCommand(GlobalInboundInvocationHandler.java:123)
        at org.infinispan.remoting.inboundhandler.GlobalInboundInvocationHandler.handleFromCluster(GlobalInboundInvocationHandler.java:77)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processRequest(JGroupsTransport.java:1527)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1454)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1649)
        at org.jgroups.JChannel.up(JChannel.java:748)
        at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:941)
        at org.jgroups.protocols.FRAG2.up(FRAG2.java:139)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:261)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:253)
        at org.jgroups.protocols.pbcast.GMS.up(GMS.java:853)
        at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:235)
        at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:462)
        at org.jgroups.protocols.pbcast.NAKACK2.deliver(NAKACK2.java:1006)
        at org.jgroups.protocols.pbcast.NAKACK2.handleMessage(NAKACK2.java:894)
        at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:679)
        at org.jgroups.protocols.VERIFY_SUSPECT2.up(VERIFY_SUSPECT2.java:105)
        at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:180)
        at org.jgroups.protocols.FD_SOCK2.up(FD_SOCK2.java:190)
        at org.jgroups.protocols.MERGE3.up(MERGE3.java:274)
        at org.jgroups.protocols.Discovery.up(Discovery.java:296)
        at org.jgroups.stack.Protocol.up(Protocol.java:360)
        at org.jgroups.protocols.TP.passMessageUp(TP.java:1230)
        at org.jgroups.util.SubmitToThreadPool$SingleMessageHandler.run(SubmitToThreadPool.java:95)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.ClassCastException: class org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent cannot be cast to class org.keycloak.models.cache.infinispan.events.UserCacheInvalidationEvent (org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent and org.keycloak.models.cache.infinispan.events.UserCacheInvalidationEvent are in unnamed module of loader io.quarkus.bootstrap.runner.RunnerClassLoader @7e6cbb7a)
        at org.keycloak.models.cache.infinispan.UserCacheManager.addInvalidationsFromEvent(UserCacheManager.java:96)
        at org.keycloak.models.cache.infinispan.CacheManager.invalidationEventReceived(CacheManager.java:218)
        at org.keycloak.models.cache.infinispan.InfinispanUserCacheProviderFactory.lambda$lazyInit$0(InfinispanUserCacheProviderFactory.java:65)
        at org.keycloak.cluster.ClusterEvent.accept(ClusterEvent.java:29)
        at org.keycloak.cluster.ClusterEvent.accept(ClusterEvent.java:25)
        at java.base/java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891)
        at org.keycloak.cluster.infinispan.InfinispanNotificationsManager.eventReceived(InfinispanNotificationsManager.java:290)
        at org.keycloak.cluster.infinispan.InfinispanNotificationsManager$CacheEntryListener.cacheEntryCreated(InfinispanNotificationsManager.java:187)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.infinispan.notifications.impl.AbstractListenerImpl$ListenerInvocationImpl.lambda$invoke$1(AbstractListenerImpl.java:431)
        ... 74 more

2024-10-10 11:36:27,748 ERROR [org.infinispan.interceptors.impl.InvocationContextInterceptor] (jgroups-17,keycloak26-keycloakx-0-47006) ISPN000136: Error executing command PutKeyValueCommand on Cache 'work', writing keys [8a0e214a-b415-4c92-9a47-ce413c44444f]: org.infinispan.commons.CacheListenerException: ISPN000280: Caught exception [java.lang.ClassCastException] while invoking method [public void org.keycloak.cluster.infinispan.InfinispanNotificationsManager$CacheEntryListener.cacheEntryCreated(org.infinispan.notifications.cachelistener.event.CacheEntryCreatedEvent)] on listener instance: org.keycloak.cluster.infinispan.InfinispanNotificationsManager$CacheEntryListener@1e03f0cb
        at org.infinispan.notifications.impl.AbstractListenerImpl$ListenerInvocationImpl.lambda$invoke$1(AbstractListenerImpl.java:437)
        at org.infinispan.notifications.impl.AbstractListenerImpl$ListenerInvocationImpl.invoke(AbstractListenerImpl.java:457)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.doRealInvocation(CacheNotifierImpl.java:1793)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.doRealInvocation(CacheNotifierImpl.java:1789)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.invokeNoChecks(CacheNotifierImpl.java:1784)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.invoke(CacheNotifierImpl.java:1757)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl.doNotifyCreated(CacheNotifierImpl.java:428)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl.notifyCacheEntryCreated(CacheNotifierImpl.java:410)
        at org.infinispan.notifications.cachelistener.NotifyHelper.entryCommitted(NotifyHelper.java:63)
        at org.infinispan.interceptors.locking.ClusteringDependentLogic$ReplicationLogic.commitSingleEntry(ClusteringDependentLogic.java:505)
        at org.infinispan.interceptors.locking.ClusteringDependentLogic$AbstractClusteringDependentLogic.commitEntry(ClusteringDependentLogic.java:239)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.commitContextEntry(EntryWrappingInterceptor.java:611)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.commitEntryIfNeeded(EntryWrappingInterceptor.java:884)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.commitContextEntries(EntryWrappingInterceptor.java:591)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.applyChanges(EntryWrappingInterceptor.java:685)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.applyAndFixVersion(EntryWrappingInterceptor.java:737)
        at org.infinispan.interceptors.InvocationSuccessFunction.apply(InvocationSuccessFunction.java:25)
        at org.infinispan.interceptors.impl.QueueAsyncInvocationStage.invokeQueuedHandlers(QueueAsyncInvocationStage.java:124)
        at org.infinispan.interceptors.impl.QueueAsyncInvocationStage.accept(QueueAsyncInvocationStage.java:87)
        at org.infinispan.interceptors.impl.QueueAsyncInvocationStage.accept(QueueAsyncInvocationStage.java:33)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
        at org.infinispan.remoting.transport.AbstractRequest.complete(AbstractRequest.java:69)
        at org.infinispan.remoting.transport.impl.MultiTargetRequest.onResponse(MultiTargetRequest.java:111)
        at org.infinispan.remoting.transport.impl.RequestRepository.addResponse(RequestRepository.java:51)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processResponse(JGroupsTransport.java:1556)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1457)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1649)
        at org.jgroups.JChannel.up(JChannel.java:748)
        at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:941)
        at org.jgroups.protocols.FRAG2.up(FRAG2.java:139)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:245)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:245)
        at org.jgroups.protocols.pbcast.GMS.up(GMS.java:853)
        at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:235)
        at org.jgroups.protocols.UNICAST3.deliverMessage(UNICAST3.java:1144)
        at org.jgroups.protocols.UNICAST3.addMessage(UNICAST3.java:880)
        at org.jgroups.protocols.UNICAST3.handleDataReceived(UNICAST3.java:862)
        at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:474)
        at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:669)
        at org.jgroups.protocols.VERIFY_SUSPECT2.up(VERIFY_SUSPECT2.java:105)
        at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:180)
        at org.jgroups.protocols.FD_SOCK2.up(FD_SOCK2.java:190)
        at org.jgroups.protocols.MERGE3.up(MERGE3.java:274)
        at org.jgroups.protocols.Discovery.up(Discovery.java:296)
        at org.jgroups.stack.Protocol.up(Protocol.java:360)
        at org.jgroups.protocols.TP.passMessageUp(TP.java:1230)
        at org.jgroups.util.SubmitToThreadPool$SingleMessageHandler.run(SubmitToThreadPool.java:95)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.ClassCastException: class org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent cannot be cast to class org.keycloak.models.cache.infinispan.events.UserCacheInvalidationEvent (org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent and org.keycloak.models.cache.infinispan.events.UserCacheInvalidationEvent are in unnamed module of loader io.quarkus.bootstrap.runner.RunnerClassLoader @7e6cbb7a)
        at org.keycloak.models.cache.infinispan.UserCacheManager.addInvalidationsFromEvent(UserCacheManager.java:96)
        at org.keycloak.models.cache.infinispan.CacheManager.invalidationEventReceived(CacheManager.java:218)
        at org.keycloak.models.cache.infinispan.InfinispanUserCacheProviderFactory.lambda$lazyInit$0(InfinispanUserCacheProviderFactory.java:65)
        at org.keycloak.cluster.ClusterEvent.accept(ClusterEvent.java:29)
        at org.keycloak.cluster.ClusterEvent.accept(ClusterEvent.java:25)
        at java.base/java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891)
        at org.keycloak.cluster.infinispan.InfinispanNotificationsManager.eventReceived(InfinispanNotificationsManager.java:290)
        at org.keycloak.cluster.infinispan.InfinispanNotificationsManager$CacheEntryListener.cacheEntryCreated(InfinispanNotificationsManager.java:187)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.infinispan.notifications.impl.AbstractListenerImpl$ListenerInvocationImpl.lambda$invoke$1(AbstractListenerImpl.java:431)
        ... 52 more

2024-10-10 11:36:27,748 WARN  [org.infinispan.CLUSTER] (jgroups-17,keycloak26-keycloakx-0-47006) ISPN000071: Caught exception when handling command SingleRpcCommand{cacheName='work', command=PutKeyValueCommand{key=8a0e214a-b415-4c92-9a47-ce413c44444f, value=WrapperClusterEvent [ eventKey=USER_INVALIDATION_EVENTS, sender=keycloak26-keycloakx-1-37804, senderSite=null, delegateEvents=[UserUpdatedEvent [ userId=572eec66-c00f-4c89-91b8-d84997e27835, username=matthew@mapli.co.za, email=matthew@mapli.co.za ], org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent@bf1178c6, org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent@2a2fce4f] ], flags=[SKIP_CACHE_STORE, IGNORE_RETURN_VALUES], commandInvocationId=CommandInvocation:keycloak26-keycloakx-1-37804:22, putIfAbsent=false, returnEntry=false, valueMatcher=MATCH_ALWAYS, metadata=EmbeddedExpirableMetadata{version=null, lifespan=120000, maxIdle=-1}, internalMetadata=null, successful=true, topologyId=15, segment=139}}: org.infinispan.commons.CacheListenerException: ISPN000280: Caught exception [java.lang.ClassCastException] while invoking method [public void org.keycloak.cluster.infinispan.InfinispanNotificationsManager$CacheEntryListener.cacheEntryCreated(org.infinispan.notifications.cachelistener.event.CacheEntryCreatedEvent)] on listener instance: org.keycloak.cluster.infinispan.InfinispanNotificationsManager$CacheEntryListener@1e03f0cb
        at org.infinispan.notifications.impl.AbstractListenerImpl$ListenerInvocationImpl.lambda$invoke$1(AbstractListenerImpl.java:437)
        at org.infinispan.notifications.impl.AbstractListenerImpl$ListenerInvocationImpl.invoke(AbstractListenerImpl.java:457)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.doRealInvocation(CacheNotifierImpl.java:1793)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.doRealInvocation(CacheNotifierImpl.java:1789)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.invokeNoChecks(CacheNotifierImpl.java:1784)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl$BaseCacheEntryListenerInvocation.invoke(CacheNotifierImpl.java:1757)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl.doNotifyCreated(CacheNotifierImpl.java:428)
        at org.infinispan.notifications.cachelistener.CacheNotifierImpl.notifyCacheEntryCreated(CacheNotifierImpl.java:410)
        at org.infinispan.notifications.cachelistener.NotifyHelper.entryCommitted(NotifyHelper.java:63)
        at org.infinispan.interceptors.locking.ClusteringDependentLogic$ReplicationLogic.commitSingleEntry(ClusteringDependentLogic.java:505)
        at org.infinispan.interceptors.locking.ClusteringDependentLogic$AbstractClusteringDependentLogic.commitEntry(ClusteringDependentLogic.java:239)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.commitContextEntry(EntryWrappingInterceptor.java:611)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.commitEntryIfNeeded(EntryWrappingInterceptor.java:884)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.commitContextEntries(EntryWrappingInterceptor.java:591)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.applyChanges(EntryWrappingInterceptor.java:685)
        at org.infinispan.interceptors.impl.EntryWrappingInterceptor.applyAndFixVersion(EntryWrappingInterceptor.java:737)
        at org.infinispan.interceptors.InvocationSuccessFunction.apply(InvocationSuccessFunction.java:25)
        at org.infinispan.interceptors.impl.QueueAsyncInvocationStage.invokeQueuedHandlers(QueueAsyncInvocationStage.java:124)
        at org.infinispan.interceptors.impl.QueueAsyncInvocationStage.accept(QueueAsyncInvocationStage.java:87)
        at org.infinispan.interceptors.impl.QueueAsyncInvocationStage.accept(QueueAsyncInvocationStage.java:33)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
        at org.infinispan.remoting.transport.AbstractRequest.complete(AbstractRequest.java:69)
        at org.infinispan.remoting.transport.impl.MultiTargetRequest.onResponse(MultiTargetRequest.java:111)
        at org.infinispan.remoting.transport.impl.RequestRepository.addResponse(RequestRepository.java:51)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processResponse(JGroupsTransport.java:1556)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1457)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1649)
        at org.jgroups.JChannel.up(JChannel.java:748)
        at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:941)
        at org.jgroups.protocols.FRAG2.up(FRAG2.java:139)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:245)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:245)
        at org.jgroups.protocols.pbcast.GMS.up(GMS.java:853)
        at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:235)
        at org.jgroups.protocols.UNICAST3.deliverMessage(UNICAST3.java:1144)
        at org.jgroups.protocols.UNICAST3.addMessage(UNICAST3.java:880)
        at org.jgroups.protocols.UNICAST3.handleDataReceived(UNICAST3.java:862)
        at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:474)
        at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:669)
        at org.jgroups.protocols.VERIFY_SUSPECT2.up(VERIFY_SUSPECT2.java:105)
        at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:180)
        at org.jgroups.protocols.FD_SOCK2.up(FD_SOCK2.java:190)
        at org.jgroups.protocols.MERGE3.up(MERGE3.java:274)
        at org.jgroups.protocols.Discovery.up(Discovery.java:296)
        at org.jgroups.stack.Protocol.up(Protocol.java:360)
        at org.jgroups.protocols.TP.passMessageUp(TP.java:1230)
        at org.jgroups.util.SubmitToThreadPool$SingleMessageHandler.run(SubmitToThreadPool.java:95)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.ClassCastException: class org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent cannot be cast to class org.keycloak.models.cache.infinispan.events.UserCacheInvalidationEvent (org.keycloak.models.cache.infinispan.events.CacheKeyInvalidatedEvent and org.keycloak.models.cache.infinispan.events.UserCacheInvalidationEvent are in unnamed module of loader io.quarkus.bootstrap.runner.RunnerClassLoader @7e6cbb7a)
        at org.keycloak.models.cache.infinispan.UserCacheManager.addInvalidationsFromEvent(UserCacheManager.java:96)
        at org.keycloak.models.cache.infinispan.CacheManager.invalidationEventReceived(CacheManager.java:218)
        at org.keycloak.models.cache.infinispan.InfinispanUserCacheProviderFactory.lambda$lazyInit$0(InfinispanUserCacheProviderFactory.java:65)
        at org.keycloak.cluster.ClusterEvent.accept(ClusterEvent.java:29)
        at org.keycloak.cluster.ClusterEvent.accept(ClusterEvent.java:25)
        at java.base/java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891)
        at org.keycloak.cluster.infinispan.InfinispanNotificationsManager.eventReceived(InfinispanNotificationsManager.java:290)
        at org.keycloak.cluster.infinispan.InfinispanNotificationsManager$CacheEntryListener.cacheEntryCreated(InfinispanNotificationsManager.java:187)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.infinispan.notifications.impl.AbstractListenerImpl$ListenerInvocationImpl.lambda$invoke$1(AbstractListenerImpl.java:431)
        ... 52 more

--- a/model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/events/CacheKeyInvalidatedEvent.java
+++ b/model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/events/CacheKeyInvalidatedEvent.java
@@ -23,9 +23,10 @@ import org.infinispan.protostream.annotations.ProtoFactory;
 import org.infinispan.protostream.annotations.ProtoTypeId;
 import org.keycloak.marshalling.Marshalling;
 import org.keycloak.models.cache.infinispan.RealmCacheManager;
+import org.keycloak.models.cache.infinispan.UserCacheManager;
 
 @ProtoTypeId(Marshalling.CACHE_KEY_INVALIDATION_EVENT)
-public class CacheKeyInvalidatedEvent extends InvalidationEvent implements RealmCacheInvalidationEvent {
+public class CacheKeyInvalidatedEvent extends InvalidationEvent implements RealmCacheInvalidationEvent, UserCacheInvalidationEvent {
 
     @ProtoFactory
     public CacheKeyInvalidatedEvent(String id) {
@@ -35,5 +36,10 @@ public class CacheKeyInvalidatedEvent extends InvalidationEvent implements Realm
     @Override
     public void addInvalidations(RealmCacheManager realmCache, Set<String> invalidations) {
         realmCache.invalidateCacheKey(getId(), invalidations);
+    }
+
+    @Override
+    public void addInvalidations(UserCacheManager userCache, Set<String> invalidations) {
+        userCache.invalidateCacheKey(getId(), invalidations);
     }
 }

